aptos-move/framework/aptos-framework/sources/account.move

```diff
@@ -342,6 +349,63 @@ module aptos_framework::account {
+            sequence_number: account_resource.sequence_number,
+            recipient_address,
+        };
+
+        // verify that the `SignerCapabilityOfferProofChallenge` is correct and signed by the account owner's private key
+        if (account_scheme == ED25519_SCHEME) {
+            let pubkey = ed25519::new_unvalidated_public_key_from_bytes(account_public_key_bytes);
+            let matching_auth_key_to_public_key_bytes = ed25519::unvalidated_public_key_to_authentication_key(&pubkey);
+            assert!(account_resource.authentication_key == matching_auth_key_to_public_key_bytes, error::invalid_argument(EWRONG_CURRENT_PUBLIC_KEY));
+
+            let signer_capability_sig = ed25519::new_signature_from_bytes(signer_capability_sig_bytes);
+            assert!(ed25519::signature_verify_strict_t(&signer_capability_sig, &pubkey, signer_capability_offer_proof_challenge), error::invalid_argument(EINVALID_PROOF_OF_KNOWLEDGE));
+        } else if (account_scheme == MULTI_ED25519_SCHEME) {
+            let pubkey = multi_ed25519::new_unvalidated_public_key_from_bytes(account_public_key_bytes);
+            let matching_auth_key_to_public_key_bytes = multi_ed25519::unvalidated_public_key_to_authentication_key(&pubkey);
+            assert!(account_resource.authentication_key == matching_auth_key_to_public_key_bytes, error::invalid_argument(EWRONG_CURRENT_PUBLIC_KEY));
+
+            let signer_capability_sig = multi_ed25519::new_signature_from_bytes(signer_capability_sig_bytes);
+            assert!(multi_ed25519::signature_verify_strict_t(&signer_capability_sig, &pubkey, signer_capability_offer_proof_challenge), error::invalid_argument(EINVALID_PROOF_OF_KNOWLEDGE));
+        } else {
+            abort EINVALID_SCHEME
+        };
+
+        // update the existing signer capability offer or put in a new signer capability offer for the current account
+        if (option::is_some(&account_resource.signer_capability_offer.for)) {
+            option::swap(&mut account_resource.signer_capability_offer.for, recipient_address);
+        } else {
+            option::fill(&mut account_resource.signer_capability_offer.for, recipient_address);
+        };
+    }
+
+    /// Return a signer of the offerer, if there's an existing signer/rotation capability offer at the offerer's address
+    public fun create_authorized_signer(account: &signer, offerer_address: address) : signer acquires Account {
+        assert!(exists_at(offerer_address), error::not_found(EACCOUNT_DOES_NOT_EXIST));
+
+        // Check if there's an existing signer capability offer from the offerer
+        let account_resource = borrow_global_mut<Account>(offerer_address);
+        let addr = signer::address_of(account);
+        assert!(option::contains(&account_resource.signer_capability_offer.for, &addr) || option::contains(&account_resource.rotation_capability_offer.for, &addr), error::not_found(EINVALID_ACCEPT_SIGNER_CAPABILITY));
```

Comment:
Uhm, I thought the signer capability was necessary to create a signer. It seems like the rotation capability suffices too though.

Why do we need two different capabilities for the same action: creating a signer? @davidiw

Reply:
Because of this, it seems like a recipient that has the capability to rotate an offerer is able not just to rotate the offerer's key but also do other things, like transact?

Reply:
i think you're right, i accidentally added it here. updating now : )